### ISSUE TITLE
Question: Can you combine VAD & MFFC into a functional block to reduce load?

### ISSUE THREAD

**Issue by StuartIanNaylor** [JuliaDSP/MFCC.jl#23]
Hi just a noob and niether C or Julia savvy but have a vague understanding of what happens with both VAD & MFFC processes.

https://github.com/linto-ai/gpu-ne10-mfcc I noticed they had dropped to use of a python routine in https://github.com/linto-ai/pyrtstools and dunno why but was searching for alternatives.

With VoiceAI it seems common to have a VAD preprocessor that feeds MFCC output as 2 seperate functions and as a noob was just curious if these could be combined with so much similar functionality.
Isn't VAD just creating an overall bin average of what a MFFC would output so the 2 can be combined?

They opted for https://github.com/MycroftAI/sonopy/blob/master/sonopy.py and https://github.com/wiseman/py-webrtcvad and not sure what was wrong with the neon optimised version they had.

PS about all I know about Julia is that it could become a replacement for Python... eventually :)
Its supposed to be fast in the respect to the above what sort of speed increases and load reduction are likely.

Its only curiosity as the repos posted I am not a contributor but as above wondered with the simularity why they are not combined?

Its only curiosity, so can not presure for an answer, but just wondering how it would also relate in performce to native Neon code as in.
https://projectne10.github.io/Ne10/doc/group__groupDSPs.html

**Comment by davidavdav** [comment]
Hi, since you're starting in this field, please refer to Speech Activity Detection as SAD and not VAD, despite that a lot of people call this VAD—but that is a confusing terminology.  If you do that from the beginning, you'll get used to using SAD rather than VAD.

Usually, you do SAD as a separate task.  SAD per se is a very hard but non-sexy task.  With clean speech signals there is a plethora of ways to do this easily and quickly, e.g., based on a per-frame energy, but things get harder when there is noise of unknown source in the signal at an unknown and/or varying signal to noise ratio. 

For many applications, there is no real information in the silent (i.e. non-speech) parts of the signal, so in further processing frames without speech can be discarded in further processing.

**Comment by StuartIanNaylor** [comment]
Lets be SAD then :)

I was just actually wondering about the FFT implementation of SAD which again I will get the terminology wrong but the frequency bands sometimes are reffered to as bins that are tallied to give a sort of RMS of the overall voice spectrum.
What is overall voice spectrum and including fundementals and all that, lets not go there.

With in rough terminology you are making an intensity 'map' of the same frequency bands.

The question was mainly in terms of code with the simularity of FFT function could the SAD and MFFC be run by the same code with both processes concurrently?
Both are sort of non sexy and the math and terminology behind them for the likes of me is extremely hard but are they not prectically the same.
The difference being SAD is working on a singular frame whilst the MFFC is working on a window of frames from SAD no silence to SAD silence again.

I am bound to through in a VAD as so many frameworks call it that but think I managed teh SAD quite well :)

The noise of unknown source is a problem but outside the complexity and needs as its fed firstly from AEC.

I have been using on the good old Pi https://github.com/voice-engine/ec which is just SpeexDSP as Alsa plugins Speex Echo or Webrtc Aec via PulseAudio seem to mangle the source audio you want quite badly.
The above works quite well and could be said to be a Echo attenutor rather than cancellor but its good enough with the software alternatives I have.

So with the SAD processing  there is noise but speech input should be decibels louder so the SAD value should of that noise shouldn't trigger the threshold.
But I am straying away from the commonality between the 2 functions and rather have something like webrtcvad creating a wav or stream for MFCC could they not just be part of the same process?

**Comment by StuartIanNaylor** [comment]
Holy hell just been reading https://labrosa.ee.columbia.edu/matlab/rastamat/ and what you have there.
I had only flicked through your read.me as was wondering if MFCC could be processed in the same way as feacalc(:energy) apart it creates and envelope similar to ALC from 'silence' to 'silence' for sentance part splitting. 

I didn't know what I was looking at but there is a whole rake of reasonably sized VoiceAI projects that could benefit by implementing what seems to be the grand daddy of MFCC libs.

With the MFCC is it paramterised to be able to raise the log-mel-amplitudes to a suitable power (around 2 or 3) before taking the DCT?

I have posted your repo off to https://github.com/linto-ai/pyrtstools as I really like what they are doing with datasets in providing much easier routes for collection and model creations.
https://doc.linto.ai/#/repos blackbox models are not acceptable and the dataset is the code, so hats off to them for that.
Mozilla common voice provide world class dataset collection and validation and then just present it in one huge ugly tar!?

Also to https://github.com/synesthesiam who is godlike in providing opensource voiceAI projects that are usable and a learning platform for the likes of me. AKA Rhasspy and Voice2Json and quite a few others.
